overlord/servicestate/quotastate.go

```diff
@@ -0,0 +1,548 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package servicestate
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/snapcore/snapd/gadget/quantity"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/progress"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/quota"
+	"github.com/snapcore/snapd/snapdenv"
+	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/systemd"
+	"github.com/snapcore/snapd/timings"
+	"github.com/snapcore/snapd/wrappers"
+	tomb "gopkg.in/tomb.v2"
+)
+
+// QuotaControlAction is the serialized representation of a quota group
+// modification that lives in a task.
+type QuotaControlAction struct {
+	// QuotaName is the name of the quota group being controlled.
+	QuotaName string `json:"quota-name"`
+
+	// Action is the action being taken on the quota group. It can be either
+	// "create", "update", or "remove".
+	Action string `json:"action"`
+
+	// AddSnaps is the set of snaps to add to the quota group, valid for either
+	// the "update" or the "create" actions.
+	AddSnaps []string `json:"snaps"`
+
+	// MemoryLimit is the memory limit for the quota group being controlled,
+	// either the initial limit the group is created with for the "create"
+	// action, or if non-zero for the "update" the memory limit, then the new
+	// value to be set.
+	MemoryLimit quantity.Size
+
+	// ParentName is the name of the parent for the quota group if it is being
+	// created. Eventually this could be used with the "update" action to
+	// support moving quota groups from one parent to another, but that is
+	// currently not supported.
+	ParentName string
+}
+
+func (m *ServiceManager) doQuotaControl(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	defer perfTimings.Save(st)
+
+	meter := snapstate.NewTaskProgressAdapterUnlocked(t)
+
+	qc := &QuotaControlAction{}
+	err := t.Get("quota-control-action", qc)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot get quota-control-action: %v", err)
+	}
+
+	allGrps, err := AllQuotas(st)
+	if err != nil {
+		return err
+	}
+
+	args := quotaControlArgs{
+		St:          st,
+		AllGrps:     allGrps,
+		Action:      qc,
+		Meter:       meter,
+		PerfTimings: perfTimings,
+	}
+
+	switch qc.Action {
+	case "create":
+		return doCreateQuota(args)
+	case "remove":
+		return doRemoveQuota(args)
+	case "update":
+		return doUpdateQuota(args)
+	default:
+		return fmt.Errorf("unknown action %q requested", qc.Action)
+	}
```

Comment:
hmm this was not clear to me, I was following the example of doServiceControl for this, which does not set the state of the task status, is that because the things that service-control changes do not affect the state.json the way that quota groups do? I see that some of the snapstate handlers however do this and those definitely modify state.json AIUI so I guess that makes sense

Reply:
this was clarified later on